lib/OpenQA/WebAPI/Controller/Admin/Needle.pm

```diff
@@ -124,41 +124,25 @@ sub module {
 
 sub delete {
     my ($self) = @_;
-    my @removed_ids;
-    my @errors;
-    for my $needle_id (@{$self->every_param('id')}) {
-        my $needle = $self->db->resultset('Needles')->find($needle_id);
-        if (!$needle) {
-            push(
-                @errors,
-                {
-                    id           => $needle_id,
-                    display_name => $needle_id,
-                    message      => "Unable to find $needle_id"
-                });
-            next;
-        }
-
-        my $error = $needle->remove($self->current_user);
-        if ($error) {
-            push(
-                @errors,
-                {
-                    id           => $needle_id,
-                    display_name => $needle->filename,
-                    message      => $error
-                });
-        }
-        else {
-            push(@removed_ids, $needle_id);
-        }
-    }
-    $self->emit_event('openqa_needle_delete', {id => \@removed_ids});
-    $self->render(
-        json => {
-            removed_ids => \@removed_ids,
-            errors      => \@errors
-        });
+
+    $self->gru->enqueue_and_keep_track(
+        task_name        => 'delete_needles',
+        task_description => 'deleting needles',
+        task_args        => {
+            needle_ids => $self->every_param('id'),
+            user_id    => $self->current_user->id,
+        },
+        on_success => sub {
+            my ($result) = @_;
+
+            my $removed_ids = ($result->{removed_ids} //= []);
+            $self->emit_event(openqa_needle_delete => {id => $removed_ids}) if (@$removed_ids);
+            return $self->render(json => $result);
+        },
+        on_error => sub {
+            $self->render_gru_result(@_);
+        },
+    );
```

Comment:
Do you plan to add more `on_*` handlers in the future? Otherwise it would seem sensible to return a promise.

Reply:
```perl
$self->gru->enqueue_and_keep_track_p(...)->then(sub {...})->catch(sub {...});
```

Reply:
Changed it to use Mojo::Promise.